### PR TITLE
TypeScript Refactor

### PR DESCRIPTION
**Pull request by leonardoventurini**
> Had to create due to strange git conflict behavior on the older branch.

- Use SWC instead of babel
- Add TypeScript Support
- ESLint + Prettier TypeScript Support
- Mocha TypeScript Support + SWC
- Optimize Scripts

### CODE REVIEW

File: .npmignore [hunk 1]
```diff
@@ -1,14 +1,13 @@
 .*
 *.swp
 *~
-*.iml
 .*.haste_cache.*
 .DS_Store
-.idea
 npm-debug.log
-.eslintcache
```
[the remark this answers is not in the document]
Comment (leonardoventurini) [reply .npmignore:9]: Those are all already included in `.*`

File: .swcrc [hunk 2]
```diff
@@ -0,0 +1,16 @@
+{
```
[the remark this answers is not in the document]
Comment (leonardoventurini) [reply .swcrc:1]: Nope

File: src/utils/platform.ts [hunk 3]
```diff
@@ -0,0 +1,6 @@
+export const NodeVersion = parseInt(
+  process.version.match(/v(\d+)/)?.[1] as string,
+  10,
+);
+
+export const SupportsAsyncHooks = NodeVersion >= 12;
```
[the remark this answers is not in the document]
Comment (leonardoventurini) [reply src/utils/platform.ts:9]: Apparently `AsyncLocalStorage` is not, we get that AsyncLocalStorage is not a constructor.

[the remark this answers is not in the document]
Comment (zodern) [reply src/utils/platform.ts:9]: Maybe it should be named SupportsAsyncLocalStorage?

Async Local Storage was added in Node 12.17, so you would also want to check the minor version. Or you could have it instead check if async_hooks exports AsyncLocalStorage.

[the remark this answers is not in the document]
Comment (leonardoventurini) [reply src/utils/platform.ts:9]: That makes sense, thanks

File: tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+    "strict": true,
+    "allowJs": true,
+    "allowSyntheticDefaultImports": true,
+    "allowUnreachableCode": true,
+    "declaration": true,
+    "esModuleInterop": true,
+    "incremental": true,
+    "module": "CommonJS",
+    "moduleResolution": "node",
+    "noImplicitAny": false,
```
[the remark this answers is not in the document]
Comment (leonardoventurini) [reply tsconfig.json]: No, I reenabled it and adjusted the missing types.

File: src/utils/websocket-utils.ts [hunk 5]
```diff
@@ -35,12 +35,14 @@ export function connectWebSocket(url, headers, onMessage) {
     ws.on(WebSocketEvent.OPEN, () => {
       // Need to remove the handlers, otherwise they will
       // be called again in normal operation
-      ws.off(WebSocketEvent.CLOSE, errorHandler);
-      ws.off(WebSocketEvent.ERROR, errorHandler);
+      ws.off?.(WebSocketEvent.CLOSE, errorHandler);
+      ws.off?.(WebSocketEvent.ERROR, errorHandler);
+      ws.removeEventListener?.(WebSocketEvent.CLOSE, errorHandler);
+      ws.removeEventListener?.(WebSocketEvent.ERROR, errorHandler);
```
[the remark this answers is not in the document]
Comment (leonardoventurini) [reply src/utils/websocket-utils.ts:41]: That's coming from the older implementation which used `ws` I think, which seemed to not have one of them depending on the version. Anyways, will deduplicate.

**Pull request merged**